Mathlib/AlgebraicGeometry/Cover/Over.lean

```diff
@@ -0,0 +1,209 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten
+-/
+import Mathlib.AlgebraicGeometry.Pullbacks
+import Mathlib.CategoryTheory.Limits.Constructions.Over.Basic
+import Mathlib.CategoryTheory.MorphismProperty.Comma
+import Mathlib.CategoryTheory.Limits.MorphismProperty
+
+/-!
+
+# Covers of schemes over a base
+
+In this file we define the typeclass `Cover.Over`. For a cover `𝒰` of an `S`-scheme `X`,
+the datum `𝒰.Over S` contains `S`-scheme structures on the components of `𝒰` and asserts
+that the component maps are morphisms of `S`-schemes.
+
+We provide instances of `𝒰.Over S` for standard constructions on covers.
+
+-/
+
+universe v u
+
+noncomputable section
+
+open CategoryTheory Limits
+
+namespace AlgebraicGeometry.Scheme
+
+variable {P : MorphismProperty Scheme.{u}} (S : Scheme.{u})
+
+/-- Bundle an `S`-scheme with `P` into an object of `P.Over ⊤ S`. -/
+abbrev asOverProp (X : Scheme.{u}) (S : Scheme.{u}) [X.Over S] (h : P (X ↘ S)) : P.Over ⊤ S :=
+  ⟨X.asOver S, h⟩
+
+/-- Bundle an `S`-morphism of `S`-scheme with `P` into a morphism in `P.Over ⊤ S`. -/
+abbrev Hom.asOverProp {X Y : Scheme.{u}} (f : X.Hom Y) (S : Scheme.{u}) [X.Over S] [Y.Over S]
+    [f.IsOver S] {hX : P (X ↘ S)} {hY : P (Y ↘ S)} : X.asOverProp S hX ⟶ Y.asOverProp S hY :=
+  ⟨f.asOver S, trivial, trivial⟩
+
+/-- A `P`-cover of a scheme `X` over `S` is a cover, where the components are over `S` and the
+component maps commute with the structure morphisms. -/
+protected class Cover.Over {P : MorphismProperty Scheme.{u}} {X : Scheme.{u}} [X.Over S]
+    (𝒰 : X.Cover P) where
+  over (j : 𝒰.J) : (𝒰.obj j).Over S := by infer_instance
+  isOver_map (j : 𝒰.J) : (𝒰.map j).IsOver S := by infer_instance
+
+attribute [instance] Cover.Over.over Cover.Over.isOver_map
+
+instance [P.ContainsIdentities] [P.RespectsIso] {X Y : Scheme.{u}} (f : X ⟶ Y) [X.Over S] [Y.Over S]
+    [f.IsOver S] [IsIso f] : (coverOfIsIso (P := P) f).Over S where
+  over _ := inferInstanceAs <| X.Over S
+  isOver_map _ := inferInstanceAs <| f.IsOver S
+
+section
+
+variable [P.IsStableUnderBaseChange] [IsJointlySurjectivePreserving P]
+variable {X W : Scheme.{u}} (𝒰 : X.Cover P) (f : W ⟶ X) [W.Over S] [X.Over S]
+  [𝒰.Over S] [f.IsOver S]
+
+/-- The pullback of a cover of `S`-schemes along a morphism of `S`-schemes. This is not
+definitionally equal to `AlgebraicGeometry.Scheme.Cover.pullbackCover`, as here we take
+the pullback in `Over S`, whose underlying schemes is only isomorphic but not equal to the
+pullback in `Scheme`. -/
+@[simps]
+def Cover.pullbackCoverOver : W.Cover P where
```

Comment:
Why can't we use `AlgebraicGeometry.Scheme.Cover.pullbackCover` instead?

Reply:
In #18945 I use these to define a pretopology on `Over S`, for which I need to show that pullbacks of coverings are coverings. This involves an `=` of presieves and so `Cover.pullbackCover` is not sufficient I think, because of the non def-eq issue described in the comment. (I of course only need the primed versions, but I added the others for completeness). Should I add to the docstring that these should typically not be used?

Reply:
I would also think it would be much better it we could avoid reproving 4 times that a pullback of a cover is a cover. Is not it possible to prove that if `X : I → Scheme` and `Y : I → Scheme` are isomorphic families of schemes over a certain scheme `S`, then `X` covers `S` iff `Y` does (whether this is phrased in `Scheme`, the `Over S` categories or their variants with morphism properties). Then, presumably, you may transport the property that a family is covering by using the comparison isomorphisms between the pullback in the various categories that are involved.

Reply:
Thanks! I added something similar (without indices and for surjective morphisms) and this golfs the `covers` proofs quite nicely.